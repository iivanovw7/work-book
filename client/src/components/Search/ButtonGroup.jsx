import PropTypes from 'prop-types';
import React from 'react';
import moment from 'moment';
import styled from 'styled-components';
import { colorScheme } from '../../config';
import Button from '../../elements/UI/Button';
import TagButton from '../../elements/UI/TagButton';
import * as utils from '../../utils';
/* eslint no-underscore-dangle: 0 */

const StyledPostDescription = styled.div`
	display: flex;
  flex-direction: column;
  width: 100%;
  height: 100%;
  align-content: center;
  justify-content: space-between;
  
  >div:last-of-type {
  	justify-content: flex-end;
  }
`;

const ButtonGroup = (props) => {
  const {
    theme, locale, text, query, post, handlePostClick, handleTagClick
  } = props;
  const colorIndex = utils.getRandomInt(0, colorScheme.tagBtnBgColors.length);
  const date = post.created.toString();

  return (
    <StyledPostDescription>
      <div>
        <h3 className="dates">
          {moment(date, 'x')
            .format('MMMM DD, YYYY')}
        </h3>
        <Button
          variant="primary"
          text={text.navigation.readMore[locale]}
          theme={theme}
          handleClick={(e) => {
            e.preventDefault();
            handlePostClick(post._id);
          }}
        />
      </div>
      <TagButton
        text={query}
        theme={theme}
        label={text.search.query[locale]}
        bgColor={colorScheme.tagBtnBgColors[colorIndex]}
        textColor={colorScheme.tagBtnTextColors[colorIndex]}
        handleClick={(e) => {
          e.preventDefault();
          handleTagClick();
        }}
      />
    </StyledPostDescription>
  );
};

export default ButtonGroup;

ButtonGroup.propTypes = {
  locale: PropTypes.string.isRequired,
  text: PropTypes.object.isRequired,
  theme: PropTypes.string.isRequired,
  query: PropTypes.string.isRequired,
  post: PropTypes.object.isRequired,
  handlePostClick: PropTypes.func.isRequired,
  handleTagClick: PropTypes.func.isRequired
};
