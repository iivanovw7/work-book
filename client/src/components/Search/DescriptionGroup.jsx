import PropTypes from 'prop-types';
import React from 'react';
import styled from 'styled-components';
import { colorScheme } from '../../config';
import Number from '../../elements/UI/Number';
import TagButton from '../../elements/UI/TagButton';
import * as utils from '../../utils';

const StyledBottomContainer = styled.div`
	display: flex;
	flex-direction: row;
	justify-content: space-between;
`;

const DescriptionGroup = (props) => {
	const {
		theme, query, handleTagClick, post, locale
	} = props;

	const est = utils.calculateReadingTime(post.text.length, locale);
	const colorIndex = utils.getRandomInt(0, colorScheme.tagBtnBgColors.length);

	return (
		<StyledBottomContainer>
			<Number
				variant="primary"
				value={est.minutes}
				label={est.ending}
				theme={theme}
			/>
			<div className="mobile">
				<TagButton
					text={query}
					theme={theme}
					bgColor={colorScheme.tagBtnBgColors[colorIndex]}
					textColor={colorScheme.tagBtnTextColors[colorIndex]}
					handleClick={(e) => {
						e.preventDefault();
						handleTagClick();
					}}
				/>
			</div>
		</StyledBottomContainer>
	);
};

export default DescriptionGroup;

DescriptionGroup.propTypes = {
	theme: PropTypes.string.isRequired,
	query: PropTypes.string.isRequired,
	post: PropTypes.object.isRequired,
	handleTagClick: PropTypes.func.isRequired,
	locale: PropTypes.string.isRequired
};
