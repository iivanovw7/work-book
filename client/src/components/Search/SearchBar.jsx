import PropTypes from 'prop-types';
import React, { useEffect, useState } from 'react';
import styled from 'styled-components';
import useStoreon from 'storeon/react';
import validate from 'validate.js';
import Input from '../../elements/UI/Input';
import * as utils from '../../utils';
/* eslint react/require-default-props: 0 */

const SearchBarContainer = styled.div`

`;

const SearchBar = (props) => {
  const {
    history, locale, theme, text
  } = props;
  const { dispatch } = useStoreon('search');
  const [queryValidation, setQueryValidation] = useState('');
  const [searchTerm, setSearchTerm] = useState('');
  const debouncedSearchTerm = utils.useDebounce(searchTerm, 500);

  function validateInput() {
    const validationResult = validate({
      searchTerm
    }, utils.constraintsSearch);

    if (validationResult) {
      if (validationResult.searchTerm) {
        setQueryValidation(validationResult.searchTerm[0][locale]);
      } else {
        setQueryValidation('');
      }
    } else {
      setQueryValidation('');
    }

    return validationResult;
  }

  useEffect(() => {
    if (!validateInput() && debouncedSearchTerm) {
      dispatch('search', debouncedSearchTerm);
      history.push(`/search/${debouncedSearchTerm}`);
    }
  },
  [debouncedSearchTerm]);

  return (
    <SearchBarContainer>
      <Input
        label={text.search.bar.tooltip[locale]}
        id="keyword"
        type="text"
        theme={theme}
        validate={queryValidation}
        placeholder={text.search.bar.placeholder[locale]}
        onChange={e => setSearchTerm(e.target.value)}
      />
    </SearchBarContainer>
  );
};

export default SearchBar;

SearchBar.propTypes = {
  history: PropTypes.object.isRequired,
  locale: PropTypes.string.isRequired,
  theme: PropTypes.string.isRequired,
  text: PropTypes.object.isRequired
};
