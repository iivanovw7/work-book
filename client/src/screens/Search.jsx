import PropTypes from 'prop-types';
import React from 'react';
import { withRouter } from 'react-router-dom';
import { Col, Grid, Row } from 'react-styled-flexboxgrid';
import useStoreon from 'storeon/react';
import styled, { ThemeProvider } from 'styled-components';
import SearchPostsByTag from '../queryBlocks/SearchByTag';
import SearchPostsByKeyword from '../queryBlocks/SearchByKeyword';
import MobileNavBar from '../components/Navigation/MobileNavbar';
import SideBar from '../components/Navigation/SideBar';
import TopBar from '../components/Navigation/TopBar';
import { backgroundColor, textColor } from '../theme';
/* eslint react/require-default-props: 0 */
/* eslint no-underscore-dangle: 0 */

const StyledSection = styled.section`
	color: ${textColor};
  background-color: ${backgroundColor};
  min-width: 100%
  
  box-sizing: content-box;
`;

const StyledSideBar = styled.aside`
	max-width: inherit;
`;

const Search = (props) => {
  const { location } = props;
  const {
    theme, locale, localizedText, search
  } = useStoreon('theme', 'locale', 'localizedText', 'search');
  const pathName = location.pathname.split('/')[2];

  return (
    <ThemeProvider theme={{ mode: theme }}>
      <StyledSection className="container">
        <section>
          <TopBar locale={locale} text={localizedText} {...props} />
        </section>
        <Grid>
          <Row>
            <Col xs={12} sm={12} md={3} lg={3}>
              <StyledSideBar>
                <SideBar
                  theme={theme}
                  locale={locale}
                  location={location}
                  text={localizedText}
                  onlyMobile={false}
                  {...props}
                />
              </StyledSideBar>
            </Col>
            <Col xs={12} sm={12} md={9} lg={9}>
              { search !== '' && (
                <SearchPostsByKeyword
                  keyword={search}
                  locale={locale}
                  text={localizedText}
                  theme={theme}
                  {...props}
                />
              )}
              { search === '' && (
                <SearchPostsByTag
                  tag={pathName}
                  locale={locale}
                  text={localizedText}
                  theme={theme}
                  {...props}
                />
              )}
            </Col>
          </Row>
        </Grid>
        <section className="mobile">
          <MobileNavBar theme={theme} locale={locale} text={localizedText} {...props} />
        </section>
      </StyledSection>
    </ThemeProvider>
  );
};

Search.propTypes = {
  history: PropTypes.object.isRequired,
  location: PropTypes.object.isRequired,
  user: PropTypes.any,
  userAccess: PropTypes.bool.isRequired
};

export default withRouter(Search);
