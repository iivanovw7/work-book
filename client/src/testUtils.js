import { createMemoryHistory } from 'history';
import * as localized from './assets/locales.json';

// Mock browser history
export const history = createMemoryHistory('/dashboard');

// Grid config object for testing
export const mappedSizes = {
	xs: 12,
	sm: 12,
	md: 12,
	lg: 11
};

export const localizedText = {
	locales: localized.locales,
	topBar: localized.topBar,
	login: localized.login,
	navigation: localized.navigation,
	pageNotFound: localized.pageNotFound
};
