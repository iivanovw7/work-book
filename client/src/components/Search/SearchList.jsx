import PropTypes from 'prop-types';
import React from 'react';
import styled from 'styled-components';
import { opacify } from 'polished';
import { fadeInTop } from '../../styles';
import { gridConfig } from '../../config';
import TextLink from '../../elements/UI/TextLink';
import ButtonGroup from './ButtonGroup';
import DescriptionGroup from './DescriptionGroup';
/* eslint no-underscore-dangle: 0 */

const SearchWrapper = styled.div`
	@media screen and (max-width: ${gridConfig.container.md}rem) {
  	flex-direction: column;
  	>div:first-of-type, >div:last-of-type {
  		width: 100%;
  	}
  }
  @media screen and (min-width: ${gridConfig.container.md}rem) {
  	flex-direction: row;
  	>div:first-of-type {
  		width: 70%;
  	}
  	>div:last-of-type {
  		width: 30%;
  	}
  }
  
  ${fadeInTop};
  display: flex;
	flex-wrap: no-wrap;
	justify-content: space-between;
	user-select: none;
	margin: 0.5em 0 0.5em 0;
	padding: 0.7em;
	border-radius: 0.2em;
	background-color: ${opacify('0.01', 'rgba(25, 0, 0, 0.1)')};
	
	>div:first-of-type {
		flex-wrap: nowrap;
    overflow-wrap: break-word;
	}
	
	>div:last-of-type {
		text-align: right;
	}
`;

const SearchList = (props) => {
  const {
    data, history, theme, locale, text, query
  } = props;

  function handlePostClick(id) {
    history.push(`/posts/${id}`);
  }

  function handleTagClick() {
    history.push(`/search/${query}`);
  }

  return (data.findPostsByTag.map(post => (
    <SearchWrapper key={post._id}>
      <div>
        <h2>
          <TextLink
            theme={theme}
            text={post.title}
            variant="primary"
            link="/"
            handleClick={(e) => {
              e.preventDefault();
              handlePostClick(post._id);
            }}
          />
        </h2>
        <h3>{post.subject}</h3>
        <DescriptionGroup
          post={post}
          theme={theme}
          query={query}
          handleTagClick={handleTagClick}
          locale={locale}
        />
      </div>
      <div className="desktop">
        <ButtonGroup
          post={post}
          text={text}
          theme={theme}
          query={query}
          handleTagClick={handleTagClick}
          handlePostClick={handlePostClick}
          locale={locale}
        />
      </div>
    </SearchWrapper>
  )));
};

export default SearchList;

SearchList.propTypes = {
  data: PropTypes.object.isRequired,
  history: PropTypes.object.isRequired,
  locale: PropTypes.string.isRequired,
  text: PropTypes.object.isRequired,
  theme: PropTypes.string.isRequired,
  query: PropTypes.string.isRequired
};
