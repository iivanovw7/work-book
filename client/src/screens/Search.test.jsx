import React from 'react';
import wait from 'waait';
import 'cross-fetch/polyfill';
import 'jsdom-global/register';
import chalk from 'chalk';
import { MemoryRouter } from 'react-router-dom';
import Search from './Search';
import * as testUtils from '../testUtils';

jest.mock('../config/apiURL', () => () => '/api');
jest.mock('storeon/react', () => () => ({
  locale: 'eng',
  theme: 'dark',
  search: ''
}));
jest.mock('../components/Navigation/MobileNavbar', () => () => (
  <div id="MobileNavbar" />
));
jest.mock('../components/Navigation/TopBar', () => () => (
  <div id="TopBar" />
));
jest.mock('../components/Navigation/SideBar', () => () => (
  <div id="SideBar" />
));
jest.mock('../graphql/SearchByTag', () => () => (
  <div id="SearchByTag" />
));
jest.mock('../components/UI/TagButton', () => () => (
  <div id="TagButton" />
));

describe(`Testing [${chalk.yellow('SEARCH')}] screen: `, () => {
  afterAll(() => {
    jest.clearAllMocks();
  });

  const props = {
    localizedText: testUtils.localizedText,
    location: {},
    user: {},
    userAccess: true
  };

  const Composition = () => (
    <MemoryRouter>
      <Search {...props} />
    </MemoryRouter>
  );

  it('Should render properly with props ', async () => {
    const component = testUtils.suppressConsoleWarnings(
      <Composition />,
      'mount',
      console
    );

    await wait(0);
    await component.update();

    expect(component.find('Search').length).toEqual(1);
    expect(component.find('#MobileNavbar').length).toEqual(1);
    expect(component.find('#TopBar').length).toEqual(1);
    expect(component.find('#SideBar').length).toEqual(1);
  });

  it('Should render and match snapshot ', async () => {
    const component = testUtils.suppressConsoleWarnings(
      <Composition />,
      'mount',
      console
    );

    await wait(0);
    await component.update();

    expect(component).toMatchSnapshot();
  });
});
