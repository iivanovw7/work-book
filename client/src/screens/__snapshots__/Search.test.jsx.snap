// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing SEARCH screen:  Should render and match snapshot  1`] = `ReactWrapper {}`;
